### PR TITLE
RFC: commitlog data segregation based on expected source behaviour

### PR DESCRIPTION
**Pull request by elcallio**
Do not merge. RFC.
A somewhat coarse, but at least partially functional version of segregation of data in CL based on source "class" (determined by caller).
Basically divides CL segment queues into two: "normal" and "low freq/volume". The idea being that we can limit the amount of forward flush request calls we generate for the low frequency set while still maintaining a somewhat healthy flow of segments in the more heavy queue. The idea being to try to reduce flushing of memtables with very low utilization (near empty), and the subsequent tiny sstables. All in the search for less compaction issues eventually. 

This test implementation only distinguishes system.paxos, based on recent LWT woes. And in local, probably not very representative runs of the C-S lwt test from #9331, the amount of flushes for this table does go down by ~75% or so. 
If this has any impact whatsoever on actual end performance is yet to be determined. 

Some other improvements, such as keeping track of which segments we've flush requested for, are in here to, and shoudl probably eventually make it to master regardless of if we decide this segmented approach is meaningful or not.

### CODE REVIEW

**Comment by avikivity** [comment]
What does it mean? do we keep two segments open concurrently, one for each class?

**Comment by avikivity** [comment]
What are "forward flush request calls"? memtable flushes? or segment file flushes?

File: db/commitlog/commitlog.cc [hunk 1]
```diff
@@ -431,7 +432,14 @@ class db::commitlog::segment_manager : public ::enable_shared_from_this<segment_
...
     segment_id_type _ids = 0;
-    std::vector<sseg_ptr> _segments;
+
+    static constexpr size_t max_data_classes = size_t(data_class::max);
+
+    template<typename T>
+    using per_data_class = std::array<T, max_data_classes>;
+
+    per_data_class<std::vector<sseg_ptr>> _segments;    
+    per_data_class<std::optional<shared_future<with_clock<db::timeout_clock>>>> _segment_allocating;
```
Comment (avikivity) [inline db/commitlog/commitlog.cc:442]: I see indeed one open segment per class. I see two problems with that:

1. even more write iops, reduces write batch size and increases write amplification. If lowfreq_lowvol is indeed very low frequency, then that shouldn't matter too much, but also bounds the gain.
2. the commitlog is no longer linear. We could restore linearity perhaps by recording an index and merging the two streams.

File: table.cc [hunk 2]
```diff
@@ -1182,6 +1182,17 @@ table::make_memtable_list() {
     return make_lw_shared<memtable_list>(std::move(seal), std::move(get_schema), _config.dirty_memory_manager, _stats, _config.memory_compaction_scheduling_group);
 }
 
+static db::data_class class_for(schema_ptr s) {
+    // TODO: an actual good way to do this check.
+    // Cannot check paxos() pointer equivalence, since the schema
+    // we get here is versioned and replaced from that initial one.
+    if (s->ks_name() == db::system_keyspace::NAME && s->cf_name() == db::system_keyspace::PAXOS) {
+        return db::data_class::lowfreq_lowvol; 
+    }
```
Comment (avikivity) [inline table.cc:1191]: Why is this lowfreq_lowvol? In some workloads it could be the most active table.

**Comment by avikivity** [comment]
I think the thing with PAXOS being near empty is that it sees a delete for every insert, so even if it has a large commitlog footprint, it has a small memtable/sstable footprint. So small sstables from these flushes are expected. What's important is to flush such a memtable when it has a large commitlog footprint, and not when it has a small commitlog footprint, so we get few such sstables (flush old memtables, not young ones).

My intuition says that the problem is lack of feedback. Commitlog issues a memtable flush, but it takes time for a response, and until the response comes we keep issuing more and more flushes, ending up with lots of small sstables.

I think we need something around this lines:

1. define "segment age" as 0 for the currently allocating segment, and 1 for the oldest segment that will cause us to exceed the memory limit.
2. whenever a segment reaches age 0.5, queue all memtables associated with it for a flush (but flush one at a time)
3. adjust the shares of the memtable_to_sstable class to be at least max_age*1000, if age > 0.5


Point 3 ensures that if commitlog reaches half occupancy, we flush with urgency, and that urgency increases the closer we get to running out of space.

**Comment by elcallio** [comment]
> 
> 
> What does it mean? do we keep two segments open concurrently, one for each class?

Yes. Or actually one full queue (open and closed segments) per class.

**Comment by elcallio** [comment]
> 
> 
> What are "forward flush request calls"? memtable flushes? or segment file flushes?

Calls from CL to client requesting client release data, so yes, memtable flush.

File: db/commitlog/commitlog.cc (hunk 1 above)
Comment (elcallio) [reply db/commitlog/commitlog.cc:442]: How does it increase write ops? It increases memory buffer footprint, and potentially disk footprint as well, as we might have so low utilization in one queue that a segment might stay alive for a long time. But neither seems likely to really impact anything severely. Actual writes to disk should be the same (except maybe in the batch mode case, where we might increase # write ops per bytes written as we might fail to sneak in as much data in buffers. This needs to be investigated, but my spider sense tells me it too is negligible.

If by linear you mean RP:s, they only need to be monotonic per CF_ID, i.e. table. Which is why this is a designator per table. As long as we put data from a single table consequently in  either queue, we should be fine. Truncation etc is recorded per table. Btw, segments will be "linear" in creation still, they will just not be discarded and reused in order in-between queues.

File: table.cc (hunk 2 above)
Comment (elcallio) [reply table.cc:1191]: Again, RFC. In the benchmark in question I observed a very low utilization of paxos memtable in most memtable flushes. The flush happens because CL reaches threshold. And indeed, the data table corresponding to the paxos table entries generated is quite full. But because we intermix data, we end up causing writes of very small sstables. 

The idea here being that while paxos might not be low frequency, it most likely has a low volume (as in actual payload). But again, needs to be verified. This is POC showing I could reduce the small sstable flushes by 75% or so for at least paxos in this benchmark. There are also small flushes for views, but that is for next step if any.

**Comment by elcallio** [comment]
> 
> 
> I think the thing with PAXOS being near empty is that it sees a delete for every insert, so even if it has a large commitlog footprint, it has a small memtable/sstable footprint. So small sstables from these flushes are expected. What's important is to flush such a memtable when it has a large commitlog footprint, and not when it has a small commitlog footprint, so we get few such sstables (flush old memtables, not young ones).
> 
> My intuition says that the problem is lack of feedback. Commitlog issues a memtable flush, but it takes time for a response, and until the response comes we keep issuing more and more flushes, ending up with lots of small sstables.
> 
> I think we need something around this lines:
> 
>     1. define "segment age" as 0 for the currently allocating segment, and 1 for the oldest segment that will cause us to exceed the memory limit.
> 
>     2. whenever a segment reaches age 0.5, queue all memtables associated with it for a flush (but flush one at a time)
> 
>     3. adjust the shares of the memtable_to_sstable class to be at least max_age*1000, if age > 0.5
> 
> 
> Point 3 ensures that if commitlog reaches half occupancy, we flush with urgency, and that urgency increases the closer we get to running out of space.

Maybe. But the case for segregation could be equally valid for example for view tables, iff the view is a very small subset of columns in the original for example. 
Or it could just be pointless complication. Again, RFC. 
I put in a patch to actually deal with repeated flushes for same ID/RP:s when hitting thresholds. This should help somewhat with "double flushing", _but_ because again RP:s are monotonic, and memtables still use RP:s to mark where they are visa-vi should it flush to sstable, there is still a case for separating data that has different access patterns. 
We only send flush requests to memtables for segments that are finished (full), but if a memtable holds data from both old and new, it will still need to flush everything, regardless of ratio.

File: table.cc (hunk 2 above)
Comment (avikivity) [reply table.cc:1191]: I think it may have low memory utilization, but high cpu/commitlog utilization due to being mostly overwrite.

Comment (elcallio) [reply table.cc:1191]: Maybe. I did not see it being overly full, but not empty either. It is not very memory heavy though (small data). I did as mentioned however manage to reduce the flushing of it quite a bit, while maintaining CL throughput/free segments mostly via the "normal" queue. If it is worth anything is debatable. And more so in anything resembling a "real" scenario (not my cramped test env).

**Comment by kostja** [comment]
As much as I like this direction in general, I think when applied to system.paxos we are re-inventing compaction here. Have we tried setting ICS compaction strategy for system.paxos? I know it's enterprise, but we could perhaps use it on system.paxos. Why are the costs of flush so high? system.paxos is just an example of a very write-intensive table, and if we shift the costs of compacting such tables to commit log segregation perhaps we're solving the problem at the wrong place?

Yet I want to stress that in general I like the idea of commit log segregation. I'd love to use this segregated commit log for raft groups instead of writing everything into system.raft table (same problem as with system.paxos, eh?). The problem is not segregating it technically and recovering from a segregated log (we could merge as many streams as we want during recovery), the problem is having *many* such segregated streams overlayed over a few files and, if we apply it to user tables, not to hand-picked system tables, developing a criteria for moving a table between different commit log streams.

To sum up:
- would it be possible to overlay different streams over some kind of block storage which is itself layered over the current segment manager? Perhaps we could move some blocks around (lsa style) to defrag the segment space?
- what is the criteria for putting things into a specific commit log group? Should we introduce a histogram with buckets for memtables of a specific memtable/commitlog amplification ratio, and have a commit log group for each histogram bucket?

**Comment by elcallio** [comment]
@kostja - to ensure RP integrity, we cannot move tables between data classes. There can be more than two data classes (though the risk of retaining partial files to long increases). The way it is constructed here, each class is basically a lower chance the CL contents for the class (completed segments) will cause a memtable flush forward call. I.e. we always exhaust classes "lower/earlier" before requesting data from a higher class. 
Technically we could perhaps move tables between classes even in runtime, we would however require keeping track of which segment in current class a table touched, and force-flush things on switchover. I think it is safer to just designate classes for different "types" of tables, based on some heuristics on their data behaviour. 

And for the record, I am 100% for solving compaction first. 
I think the patch in here that reduces duplicated flush-forward calls might be useful however. I will make a separate PR of it for non-segregated code so we can test it a bit on the big scary tests...

**Comment by kostja** [comment]
I never liked the concept of a replay position. It's a weird way to blend the physical offset in the commit log and the logical number of the log entry. I'd stop using it anywhere outside commit log and begin returning log sequence numbers, which are written explicitly to the log along with log entry contents.

**Comment by elcallio** [comment]
How would that improve anything?

**Comment by kostja** [comment]
During recovery, you would be able to open as many segment streams as you like and get a total order of writes among all segment by merging the input from all segments by log sequence numbers.

**Comment by elcallio** [comment]
The order of CL entries replayed do not matter (hello cql!). RP:s only matter for truncation purposes and during runtime, as easy water shed marks to know if a memtable is eligible for flush or not.

File: db/commitlog/commitlog.cc (hunk 1 above)
Comment (avikivity) [reply db/commitlog/commitlog.cc:442]: Let's say a shard sees 20 100-byte commits/task-quota. After the split we have 19 100-byte commits to one segment and 1 100-byte commit to another.

before: align_up(20*100, 1024) = 2048, 1 I/O
after: align_up(1*100, 1024) + align_up(19*100, 1024) = 3072, 2 I/O

1024 = xfs block size.

So we have 50% extra bytes and 100% extra I/O.

The numbers are made up but we did observe small commitlog writes (which prompted us to reduce the xfs block size and support sub-block writes).

File: table.cc (hunk 2 above)
Comment (avikivity) [reply table.cc:1191]: I still don't understand it. Even small data will still lead to memtable-induced flushing, since memtable bloat is higher than commitlog bloat. Overwrites or deletes can lead to commitlog-induced flushing. But I wouldn't characterize the paxos table as low-volume in any way.

File: db/commitlog/commitlog.cc (hunk 1 above)
Comment (elcallio) [reply db/commitlog/commitlog.cc:442]: True, but in anything resembling "normal" write pressure I more or less assume we hit the end of buffers (even if we split them) more often than the period task removes them. I.e. aligning up and wasting would be fairly rare. Can't tell for sure, as this is really only run on a single benchmark.

File: table.cc (hunk 2 above)
Comment (elcallio) [reply table.cc:1191]: I have no reason beyond observed behaviour/values when running the test from #9331. Will eventually try to re-measure (with measure being somewhat tainted by being local and scaled down) with some of the other tweaks/fixes applied. Perhaps it looks different. But I still think I will see a sizeable difference in payload data vs. paxos used to write it. 
If and where something like this should be applied is very much up for debate. I think it is worth thinking a bit about the difference in data rate (bytes/s) between tables in the write path, and the fact that singular CL for all tables do at the very least sometimes cause premature flush to sstable.
